src/tribler/core/utilities/tribler_process_manager.py

```diff
@@ -0,0 +1,175 @@
+        try:
+            return func(*args, **kwargs)
+        except sqlite3.Error as e:
+            logger.warning(f'Retrying after the error: {e.__class__.__name__}: {e}')
+            return func(*args, **kwargs)
+    new_func: func
+    return new_func
+
+
+class ProcessManager:
+    def __init__(self, root_dir: Path, process_kind: ProcessKind, creator_pid: Optional[int] = None, **other_params):
+        self.root_dir = root_dir
+        self.filename = self._get_file_name(root_dir)
+        self.current_process = TriblerProcess.current_process(process_kind, creator_pid, **other_params)
+        self.primary_process = self.atomic_get_primary_process(process_kind, self.current_process)
+
+    @classmethod
+    def _get_file_name(cls, root_dir: Path) -> Path:  # The method is added for easier testing
+        return root_dir / DB_FILENAME
+
+    def connect(self) -> sqlite3.Connection:
+        connection = sqlite3.connect(str(self.filename))
+        try:
+            connection.execute('BEGIN EXCLUSIVE TRANSACTION')
+            connection.execute(CREATE_SQL)
+            return connection
+        except:  # noqa: E722
+            connection.close()
+            raise
+
+    @contextmanager
+    def transaction(self) -> ContextManager[sqlite3.Connection]:
+        connection = None
+        try:
+            connection = self.connect()
+            yield connection
+            connection.execute('COMMIT')
+            connection.close()
+
+        except Exception as e:
+            logger.exception(f'{e.__class__.__name__}: {e}')
+            if connection:
+                connection.close()  # pragma: no cover
+            if isinstance(e, sqlite3.DatabaseError):
+                self.filename.unlink(missing_ok=True)
+            raise
+
+    @with_retry
+    def atomic_get_primary_process(self, kind: ProcessKind,
+                                  current_process: Optional[TriblerProcess] = None) -> Optional[TriblerProcess]:
+        primary_process = None
+        with self.transaction() as connection:  # pylint: disable=not-context-manager  # false Pylint alarm
+            cursor = connection.execute("""
+                SELECT * FROM processes WHERE kind = ? and "primary" = 1 ORDER BY rowid DESC LIMIT 1
+            """, [kind.value])
+            row = cursor.fetchone()
+            if row is not None:
+                previous_primary_process = TriblerProcess.from_row(row)
+                if previous_primary_process.is_running():
+                    primary_process = previous_primary_process
+                else:
+                    previous_primary_process.primary = 0
+                    previous_primary_process.save(connection)
+
+            if current_process is not None:
+                if primary_process is None:
+                    current_process.primary = 1
+                    primary_process = current_process
+                else:
+                    current_process.primary = 0
```

Comment:
I'm not sure that this line is correct.

Reply:
It was for "additional safety", but you are right that it is not necessary, and I removed it